addons/mass_mailing/models/mail_message.py

```diff
@@ -0,0 +1,52 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models, fields
+from odoo.api import depends
+from odoo.tools.misc import groupby
+
+class MailMessage(models.Model):
+    _inherit = "mail.message"
+    # ------------------------------------------------------------
+    # ADDITIONS
+    # ------------------------------------------------------------
+
+    mailing_trace_ids = fields.One2many('mailing.trace', 'mail_message_id', compute='_compute_mailing_trace_ids')
+
+    @depends('mail_ids')
+    def _compute_mailing_trace_ids(self):
+        message_trace_mapping = self._get_message_trace_mapping()
+        # browse flattened ids to do a single query
+        for message in self:
+            trace_ids = [trace.id for trace in message_trace_mapping.get(message.id, [])]
+            message.mailing_trace_ids = self.env['mailing.trace'].browse(trace_ids)
+
+    def _get_message_trace_mapping(self):
+        """Allow overrides to the compute method, adding additional results.
+
+        :return dict[int, list[int]]: Dictionary mapping message ids to trace recordsets ids
+        """
+        # need sudo to be allowed to read mail.mail
+        return dict(groupby(self.sudo().mail_ids.mailing_trace_ids.sudo(False),
+                            lambda trace_ids: trace_ids.sudo().mail_mail_id.mail_message_id.id))
+
+    # ------------------------------------------------------------
+    # OVERRIDES
+    # ------------------------------------------------------------
+    def _message_format(self, *args, **kwargs):
+        vals_list = super()._message_format(*args, **kwargs)
+        # cache
+        self.browse(vals['id'] for vals in vals_list).mailing_trace_ids
+
+        for vals in vals_list:
+            message = self.browse(vals['id'])
```

Comment:
I think it should have the same effect

```suggestion
        messages = self.browse(vals['id'] for vals in vals_list)
        for message, vals in zip(messages, vals_list):
```

Reply:
@std-odoo  I'm not sure because this assumes the ``browse`` will preserve the ordering of the vals['id']. I wouldn't want to assign the trace to the wrong dict. But if you say it's preserved then that's fine.

Reply:
I think the order is preserved :)